text/0000-alternate-registries.md

```diff
@@ -0,0 +1,120 @@
+- Feature Name: alternate-registries
+- Start Date: 2017-05-23
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Adding support for alternative crates.io servers to be used alongside the public crates.io server.
+
+# Motivation
+[motivation]: #motivation
+
+Currently cargo only has support for getting crates from a public server, this is fine for
+open source projects using Rust, however for closed source code this is problematic. Currently
+the only real option is to use Git repositories to specify the packages, but that means that
+all of the nice versioning and discoverability that cargo and crates.io provides is lost. I
+would like to change this so that it is possible to have a local crates.io server which private
+crates can be pushed to, plus still be able to make use of the public crates.io server as well.
+
+# Detailed design
+[design]: #detailed-design
+
+There are a number of different areas which will likely need to be tackled in order to fully
+support a local crates.io server in an enterprise. Below are some key areas:
+
+* Add support for alternate crates.io registries (this RFC)
+* Provide support for caching crates used on a crates.io proxy
+* Add support to crates.io to allow authentication with other OAuth providers than github
+* Support for private storage of crates on crates.io server rather than publicly available on S3
+
+Rather than trying to get everything agreed in a single RFC I would like for these to be
+dealt with as separate proposals and use this RFC as a stepping stone to be able to
+meaningfully start to use private registries.
+
+See https://github.com/rust-lang/cargo/pull/4036 for the code changes which this proposal was based on.
+
+## Cargo.toml config changes
+The following changes for Cargo.toml are proposed as part of this RFC:
+* Add an optional 'registry' field to be specified as part of the package dependencies
+* Add an optional 'registry' field to be specified in the package definition
+
+Both of the registry entries take a value of a crates.io git repository. If one is not
+provided then the default crates.io repository is assumed, this ensures that it is
+back compatible with all current crates.
+
+Below is an example of the change that we are proposing to make:
+
+```
+[package]
+name = "registry-test"
+version = "0.1.0"
+authors = ["Christopher Swindle <christopher.swindle@metaswitch.com>"]
+registry = "https://github.com/my_awesome_fork/crates.io-index"
+
+[dependencies]
+libc = { version = "*", registry = "https://github.com/my_awesome_fork/crates.io-index" }
+serde_json = "1.0.1"
+```
+
+## Changes for alternative registries for dependencies
+This boils down to a very simple change, where we previously setup the crate source for the
+crates.io registry, we now just need to check if a registry is provided, if it has the crate
+source is created using the registry URL, otherwise the crates.io server is used.
```

Comment:
I am also in favor of "per dependency" config because of the "optimize for the reader" advice. When reading Cargo.toml, I would like to know from which registry each of the deficiencies comes, and per-dep config seems to be the easiest way to achieve that.

 Adding a "default" registry would be backwards compatible, so we can skip it anyway at this stage :)